src/Client.cpp

```diff
@@ -1,245 +1,487 @@
+bool Client::connect() 
+{
+    port.open(device_name_);
+    port.set_option(asio::serial_port::baud_rate(baud_rate_));
+    port.set_option(asio::serial_port::parity(asio::serial_port::parity::none));
+    port.set_option(asio::serial_port::character_size(asio::serial_port::character_size(8)));
+    port.set_option(asio::serial_port::stop_bits(asio::serial_port::stop_bits::one));
+    return start();
+    
 }
 
-void Client::start() {
-    thread = std::thread([this]{
-        running = true;
-        while(running) { processOneMessage(); }
-    });
+bool Client::disconnect()
+{
+    port.close();
+    return stop();
 }
 
-void Client::stop() {
-    running = false;
-    pimpl->io.stop();
-    pimpl->port.close();
-    thread.join();
-}
 
-bool Client::sendData(const uint8_t id, const ByteVector &data) {
-    std::lock_guard<std::mutex> lock(mutex_send);
+bool Client::start() 
+{
+    if (!port.is_open()) return false;
+    if (running_.test_and_set()) return false;
+    thread = std::thread([this]{
+        asio::async_read_until(port, buffer, 
+            std::bind(&Client::messageReady,this,std::placeholders::_1,std::placeholders::_2),
+            std::bind(&Client::processOneMessage,this,std::placeholders::_1,std::placeholders::_2));
+        io.run();
+        
+    });
+    return true;
+}
 
-    try {
-        asio::write(pimpl->port, asio::buffer("$M<",3));              // header
-        asio::write(pimpl->port, asio::buffer({uint8_t(data.size())})); // data size
-        asio::write(pimpl->port, asio::buffer({uint8_t(id)}));          // message id
-        asio::write(pimpl->port, asio::buffer(data));                   // data
-        asio::write(pimpl->port, asio::buffer({crc(id, data)}));        // crc
-    } catch (const asio::system_error &ec) {
-        if (ec.code() == asio::error::operation_aborted) {
-            //operation_aborted error probably means the client is being closed
-            return false;
+bool Client::stop() 
+{
+    bool rc = false;
+    if (running_.test_and_set()) {
+        for (const auto& sub : subscriptions) {
+            sub.second->stop();
         }
+        
+        io.stop();
+        thread.join();
+        io.reset();
+        rc = true;
     }
-
-    return true;
+    running_.clear();
+    return rc;
 }
 
-int Client::request(msp::Request &request, const double timeout) {
-    msp::ByteVector data;
-    const int success = request_raw(uint8_t(request.id()), data, timeout);
-    if(success==1) { request.decode(data); }
-    return success;
-}
 
-int Client::request_raw(const uint8_t id, ByteVector &data, const double timeout) {
-    // send request
-    if(!sendRequest(id)) { return false; }
+bool Client::sendMessage(msp::Message& message, const double timeout)
+{
+    if (print_debug_) std::cout << "sending message - ID " << (uint32_t)message.id() << std::endl;
+    if(!sendData(message.id(), message.encode())) { 
+        std::cout << "sendData failed" << std::endl;
```

Comment:
Remove these outputs. The method already returns false and a user of this method should check the return status.